### PR TITLE
make PlanningSceneInterface add/remove CollisionObjects synchronously

### LINKED ISSUE #38: make PlanningSceneInterface add/remove CollisionObjects synchronously
I would very much like to adjust [`PlanningSceneInterface::{add/remove}CollisionObjects`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h#L86)
to use the new `ApplyPlanningScene` service such that it performs synchronous updates to the planning scene.
This is what most users expect and it is not even documented in the [tutorials](http://docs.ros.org/indigo/api/pr2_moveit_tutorials/html/planning/src/doc/move_group_interface_tutorial.html#adding-removing-objects-and-attaching-detaching-objects) that this is not the case.

At the moment these methods wrap simple `publish` calls whereas the other methods in this class query the `move_group` via services.

Thus, the current behavior is (1) inconsistent, (2) not documented, and (3) makes people add arbitrary long `sleep` calls after the function call.

**However**, ApplyPlanningScene is a MoveGroupCapability [that has to be added manually in each projects `move_group.launch`](https://github.com/ros-planning/moveit/blob/kinetic-devel/moveit_setup_assistant/templates/moveit_config_pkg_template/launch/move_group.launch#L59) and the PlanningSceneInterface has been around quite long.
Thus, simply changing the implementation would break many use-cases.
As a downward-compatible workaround I propose to check whether the service exists and either (1) call the detected service or (2) print an appropriate warning telling the user to enable ApplyPlanningScene in their move_group and afterward continue with the current publish.
With ROS-m,n, or o we could eventually remove the publish then.

### PR DESCRIPTION
**Pull request by andreaskoepf**
To fix #38 I added a helper function that calls ApplyPlanningScene service
with a fallback to publish().

@v4hn Points to discuss:
1. Was it ok to declare the `ros::ServiceClient` `mutable` to keep the ABI unmodified or should the methods be changed to non-const?
2. When using the `ApplyPlanningScene` service no diff messages are published to the "planning_scene" topic and therefore rviz does not reflect the changes. I think we definitely want to see collision objects in rviz. Could it make sense for [move_group::ApplyPlanningSceneService::applyScene()](https://github.com/ros-planning/moveit/blob/4ac0c7432d335f57aab6836cbcaaac3fccf4b6f9/moveit_ros/move_group/src/default_capabilities/apply_planning_scene_service_capability.cpp#L50) to publish 'integrated changes' to topic "planning_scene", e.g. the "planning_scene" topic could become some kind of back-channel to which only MoveIt itself publishes and everybody else (e.g. rviz) only listens?
## Todo:
- [ ] Test attach/detachObject

### CODE REVIEW

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp [hunk 1]
```diff
@@ -207,29 +209,45 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
...
 private:
 
   ros::NodeHandle node_handle_;
   ros::ServiceClient planning_scene_service_;
+  mutable ros::ServiceClient apply_planning_scene_service_;
   ros::Publisher planning_scene_diff_publisher_;
   robot_model::RobotModelConstPtr robot_model_;
+
+  // Helper function to call ApplyPlanningScene service with fallback
+  // to asynchronous processing via "planning_scene" topic.
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: nice!

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp [hunk 2]
```diff
@@ -207,29 +209,45 @@ class PlanningSceneInterface::PlanningSceneInterfaceImpl
...
     planning_scene.is_diff = true;
-    planning_scene_diff_publisher_.publish(planning_scene);
+    applyPlanningScene(planning_scene);
   }
 
 private:
 
   ros::NodeHandle node_handle_;
   ros::ServiceClient planning_scene_service_;
+  mutable ros::ServiceClient apply_planning_scene_service_;
```
Comment (davetcoleman) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: So a service member variable changes a class' state and is not const?

**Comment by davetcoleman** [comment]
+1

File: moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp (hunk 2 above)
Comment (andreaskoepf) [inline moveit_ros/planning_interface/planning_scene_interface/src/planning_scene_interface.cpp]: @davetcoleman The methods `addCollisionObjects() const` and `removeCollisionObjects() const` are both const .. but `ros::ServiceClient::call()` is not const. To use mutable was the simplest in order to not modify [planning_scene_interface/planning_scene_interface.h#L86-L89](https://github.com/Xamla/moveit/blob/d2511fb495168e0a2cafa3b989d9ad45798ddc6d/moveit_ros/planning_interface/planning_scene_interface/include/moveit/planning_scene_interface/planning_scene_interface.h#L86-L89).

**Comment by v4hn** [comment]
Please hold this off until we fixed related issues #125 and #103 .
I'll take care of this.

**Comment by rhaschke** [comment]
I think we missed an important use case of the old asynchronous publishers: Consider the situation where you have multiple `PlanningSceneMonitors` active in your setup, one - of course - in the `move_group` node, but maybe others in a node doing online collision checking, etc. In this case informing each and every PSM, would require (i) explicit knowledge about them (which probably isn't and shouldn't be available) and (ii) repeated synchronous calls to all of them.
In this situation, an asynchronous publish mechanism would be extremely helpful.

Of course you could argue, that the master `PSM` in `move_group` will eventually publish the scene update as well - but this is usually with a heavy delay. (It's throttled by default to 2Hz).

To solve the dilemma (I still think, a synchronous mechanism is useful too), I suggest to add an optional parameter (bool wait=true), that toggles between synchronous (true) and asynchronous (false) behaviour. The default should be true=synchronous.

@v4hn What do you think?

**Comment by v4hn** [comment]
@andreaskoepf: you should subscribe to `move_group/monitored_planning_scene` in RViz.

On Wed, Aug 24, 2016 at 03:21:48PM -0700, Robert Haschke wrote:
> I think we missed an important use case of the old asynchronous publishers: Consider the situation where you have multiple `PlanningSceneMonitors` active in your setup, one - of course - in the `move_group` node, but maybe others in a node doing online collision checking, etc.

Interesting point.

> In this case informing each and every PSM, would require (i) explicit knowledge about them (which probably isn't and shouldn't be available) and (ii) repeated synchronous calls to all of them.

There is no such thing as a "synchronous call to a PlanningSceneMonitor". The `ApplyPlanningScene` service is intentionally *not* part of the PSM, but part of the `move_group` instead.
So only the PSM maintained in the `move_group` can be updated synchronously.

In my opinion, you should *never ever* have more than one monitor subscribing to `/planning_scene`.
For starters you would have to make sure to get octomap updates yourself, if someone clears one octomap via the `ClearOctomap` capability, you will never see this, and one of them could simply be outdated due to network issues.
This is *exactly* why there is a `monitored_planning_scene` and/but yes, that one is throttled for performance reasons...

> To solve the dilemma (I still think, a synchronous mechanism is useful too), I suggest to add an optional parameter (bool wait=true), that toggles between synchronous (true) and asynchronous (false) behaviour. The default should be true=synchronous.

No optional parameter please, the two mechanisms are entirely different things as you just explained.
We *could* deprecate the current functions and create new `*Sync`/`*Async` versions instead to allow for the multiple-monitors-one-topic use-case.
But I'm really not sure we even *want* to support this use-case.

opinions?

**Comment by davetcoleman** [comment]
One can easily have three PSMs, right? The move_group node, a high-level application node, and Rviz...

**Comment by v4hn** [comment]
On Fri, Aug 26, 2016 at 09:30:56AM -0700, Dave Coleman wrote:
> One can easily have three PSMs, right? The move_group node, a high-level application node, and Rviz...

Sure, but do you have all of them listening to `/planning_scene` or to `move_group/monitored_planning_scene`?
We just merged a request to change the default in RViz to `move_group/monitored_planning_scene`.
Do we officially (want to) support use-cases where the PSM of the high-level application node you mention
connects to `/planning_scene` for updates? As I mentioned you will miss some information that is being incorporated
into the `PlanningScene` of the  `move_group` (and that created artifacts in RVIz until we just changed the default).

If that is the case, the only way out that I see is adding new `*Sync` versions of the relevant functions to the `planning_interface` classes
instead of replacing the publish calls by the service call under the hood..
In order to point out that the current calls are asynchronous updates, I would then also like to deprecate the current functions
and create `*ASync` functions instead.

@davetcoleman what do you think?

**Comment by davetcoleman** [comment]
Honestly its been a while since I've used the whole moveit setup on a real robot. I'm one of those planning people who lives in Rviz world with just two PSM (Rviz and my node). But when I was doing the APC last year my team enjoyed having multiple computers/Rviz's watching the PSM from different locations

**Comment by v4hn** [comment]
On Fri, Aug 26, 2016 at 12:28:32PM -0700, Dave Coleman wrote:
> But when I was doing the APC last year my team enjoyed having multiple computers/Rviz's watching the PSM from different locations

This would most likely be a case of `move_group/monitored_planning_scene` :-)

@mikeferguson what about you guys at Fetch?
Would you want to have MoveIt (somewhat) support having multiple PSM connecting to `/planning_scene`?

**Comment by v4hn** [comment]
ping @mikeferguson

**Comment by davetcoleman** [comment]
Has there been any progress on this feature?

**Comment by v4hn** [comment]
Further progress requires feedback from other developers on whether we *want* to support the use-case that multiple planning scene monitors listen to the input `/planning_scene` topic.
I believe this topic should be reserved to the monitor of the `move_group` node and my original proposal was to replace the existing function implementation.

On the contrary, if we want to support this usage, this patch would have to be changed to add *new* functions instead.
This would leave all those setups broken though that use the old interface and assume that it is implemented synchronously (the documentation does not state this is not the case).

**Comment by rhaschke** [comment]
> In my opinion, you should *never ever* have more than one monitor [the one in move_group node] subscribing to `/planning_scene`.

In this case, using a topic is the wrong concept. The topic should be replaced by a service. That will be fine for me. Either, extra functionality requiring access to the main planning scene needs to be implemented as a capability plugin to the move_group node or one needs to resort to the monitored_planning_scene topic.

**Comment by v4hn** [comment]
> In this case, using a topic is the wrong concept

I agree, but I don't think it is an option to change this, because the `planning_scene` topic is the most fundamental interface in `moveit_ros`.

By now I believe we should move the proposed functionality into new functions instead.
I'll put that on my todo list (might take some time though).

**Comment by davetcoleman** [comment]
Closing because #381 has replaced it.

**Pull request closed without merging**